Added functionality for users to embed compliance badges

sd1p opened the pull request:
Implemented issue https://github.com/bowtie-json-schema/bowtie/issues/959


https://github.com/bowtie-json-schema/bowtie/assets/43758514/bbf22e8c-1f1b-405a-907a-91eac1645acd

sd1p commented:
> Nice start! You've got some immediate conflicts. It looks like when you started you started off of an old commit, there were some changes on main that looks like you're missing. Hopefully shouldn't be hard to pull those in.
> 
> In terms of the design -- nice mockup. I think my first reaction is it's too prominent as an accordion though, this functionality is really for implementers, who are only a secondary user of that page. I think a good thing to try is to put it in the _heading_ of the top card on the page -- so a button aligned right in the heading of the card (so next to the name of the implementation) called "Bdages", which when clicked, opens a dropdown with this widget. Let me know if that makes sense!

This sounds good. I will work on it here.

sd1p commented:
Updated UI
![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/eb82520f-b8bd-4926-b681-c787a8789f2d)

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/f16dac76-222c-4649-95de-2306c8b6bf60)

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/d32b5e2d-204d-496a-99b7-552106923466)

sd1p commented:
@Julian Can I get any feedback for this PR.

Julian commented:
Overall there's a bit of a feel of "you should use this yourself for a few minutes, you'll find a few issues".

Here are a few:

* you left aligned the button instead of right aligned, so it all feels quite cramped
* you're missing badges, you just have the compliance badges. right now we have 2 categories of badges (supported versions and compliance), and then as many formats as we support there, you have 5 which seems nice -- possibly we should use pills only for selecting the formats, and maybe we simply show all the badges all at once in the dropdown, but definitely we want all of them
* there is a very unsatisfying blinking that happens when you click between the pills, the whole dropdown flashes blue
* some padding or something at the bottom of the dropdown would definitely be nice, right now it feels quite cramped at the bottom

Besides the above, I'd again suggest you spend slightly longer playing with your change before submitting it for review, but you're definitely on the right track here!

sd1p commented:
> Overall there's a bit of a feel of "you should use this yourself for a few minutes, you'll find a few issues".
> 
> Here are a few:
> 
> * you left aligned the button instead of right aligned, so it all feels quite cramped
> * you're missing badges, you just have the compliance badges. right now we have 2 categories of badges (supported versions and compliance), and then as many formats as we support there, you have 5 which seems nice -- possibly we should use pills only for selecting the formats, and maybe we simply show all the badges all at once in the dropdown, but definitely we want all of them
> * there is a very unsatisfying blinking that happens when you click between the pills, the whole dropdown flashes blue
> * some padding or something at the bottom of the dropdown would definitely be nice, right now it feels quite cramped at the bottom
> 
> Besides the above, I'd again suggest you spend slightly longer playing with your change before submitting it for review, but you're definitely on the right track here!

Thanks for the feedback. I will certainly work on component's look and feel. Initially I was too much focused on functionality of the UI. I will spend some time with it and push an update on this PR.

sd1p commented:
> * you left aligned the button instead of right aligned, so it all feels quite cramped

✅ Moved badges icon to the right.

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/a1c06ebc-9780-4811-9fa9-8d9a08255c8e)

> * you're missing badges, you just have the compliance badges. right now we have 2 categories of badges (supported versions and compliance), and then as many formats as we support there, you have 5 which seems nice -- possibly we should use pills only for selecting the formats, and maybe we simply show all the badges all at once in the dropdown, but definitely we want all of them

✅ Right now we have Version Badge + Compliance badges, added a dropdown for selecting badges type.

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/35e96232-d8da-4ddc-bf5c-81ebf91cdf87)

> * there is a very unsatisfying blinking that happens when you click between the pills, the whole dropdown flashes blue

✅ React Bootstrap dropdown was buggy, now I am using normal bootstrap classeNames for drop down

https://github.com/bowtie-json-schema/bowtie/assets/43758514/7025cd0b-1f34-44b9-b413-c02056fa48c3

>some padding or something at the bottom of the dropdown would definitely be nice, right now it feels quite cramped at the bottom

✅ Added some padding.

sd1p commented:
Made Dropdown buttons smaller 
![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/220c6597-20bd-4cec-8172-693edf1d30cf)


Please provide suggestions and feedback.

Julian commented:
Cool. We're getting closer! Thanks again. Couple of other immediate things, though I haven't looked at the code yet:

The spacing is off in the card header now, and/or it's not vertically centered. The implementation name and badges button is way closer to the top than the bottom.

> ✅ Right now we have Version Badge + Compliance badges, added a dropdown for selecting badges type.

Sorry, I was unclear. By "the dropdown" I meant the dropdown you already added when clicking the badge button. I didn't mean to add one for the badges, for that I was saying to just show all of them all at once. But I'm not sure -- maybe a dropdown as you've changed it is also reasonable, but not if we mix all of the kinds of badges together. You need somehow to preserve the structure here (that there's 1 supported version badge and N badges for compliance -- one for each dialect supported), you can't just mix in the supported versions badge with the other N all in one unlabelled list, someone will have no idea what each badge represents.

I can try to do a mockup if needed to make it super clear, but I see two options, either "show all of them" or "show a dropdown that has multiple sections, one for each type of badge" (for which we have 2 types -- a sort of Metadata badge (the versions one) and then a Compliance category (for the dialect badges).

I think another thing is there's lots of horizontal padding in the dropdown, may as well extend the box a bit. I see the font size also changes depending on what tab you click (e.g. on rST which is two lines it's smaller, whereas for markdown it's larger). Not sure how I feel about that. Think it might be better to have the size be the same.

But we're definitely getting there!

sd1p commented:
>I see the font size also changes depending on what tab you click (e.g. on rST which is two lines it's smaller, whereas for markdown it's larger).

I applied `<pre>` tag to format the rSt display text, which is causing uneven font sizes. Now I applied `<pre>` tag to all other display text make it consistent.

>You need somehow to preserve the structure here (that there's 1 supported version badge and N badges for compliance -- one for each dialect supported), you can't just mix in the supported versions badge with the other N all in one unlabeled list, someone will have no idea what each badge represents.

I will try to add sections in dropdown menu.

>I think another thing is there's lots of horizontal padding in the dropdown, may as well extend the box a bit.

I work on fixing this and make it responsive as well.

sd1p commented:
✅Added Sections to categorize badges. 

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/fce9ee31-6bf6-499b-bad3-c621b934777e)

✅Worked on look and feel of the dropdown and made it more responsive.

https://github.com/bowtie-json-schema/bowtie/assets/43758514/093d8fd8-e4ea-4343-a149-623e4f22c95c


https://github.com/bowtie-json-schema/bowtie/assets/43758514/bafa6c06-ae00-4e57-a3bd-fcfee21c5b10

sd1p commented:
Thanks for the feedback.
> > I see the font size also changes depending on what tab you click (e.g. on rST which is two lines it's smaller, whereas for markdown it's larger).
> 
> I applied `\<pre\>` tag to format the rSt display text, which is causing uneven font sizes. Now I applied `\<pre\>` tag to all other display text make it consistent.
> 
> > You need somehow to preserve the structure here (that there's 1 supported version badge and N badges for compliance -- one for each dialect supported), you can't just mix in the supported versions badge with the other N all in one unlabeled list, someone will have no idea what each badge represents.
> 
> I will try to add sections in dropdown menu.
> 
> > I think another thing is there's lots of horizontal padding in the dropdown, may as well extend the box a bit.
> 
> I work on fixing this and make it responsive as well.

sd1p commented:
@Julian Please review.

Julian commented:
@sd1p you haven't addressed

> The spacing is off in the card header now, and/or it's not vertically centered. The implementation name and badges button is way closer to the top than the bottom.

Julian commented:
There's also strange extra padding in the code blocks, e.g. I see:

![Screenshot 2024-03-04 at 4 28 25 PM](https://github.com/bowtie-json-schema/bowtie/assets/329822/df250f1f-f27b-4d72-972b-621154bf1dbf)

Again, please play with your own work for at least a few minutes looking for issues before asking for review :)

Julian commented:
Because I'm already looking at it (and sorry for the triple post) -- a dropdown menu with no label is very unusual! Probably add a title (to the whole thing) saying "Generate a Bowtie Badge", and then next to the dropdown menu a label on the left saying "Available Badges".

sd1p commented:
Thanks for the feedback I will work on improving the UI.

sd1p commented:
Fixed UI issues.

https://github.com/bowtie-json-schema/bowtie/assets/43758514/dd998206-af99-46af-819b-3f1707f0c3b5

>The spacing is off in the card header now, and/or it's not vertically centered. The implementation name and badges button is way closer to the top than the bottom.

✅ Made badge button aligned and removed bottom padding.

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/45cde504-d3a8-4f89-8c6c-eaa441058ecc)

>There's also strange extra padding in the code blocks, e.g. I see: ... https://github.com/bowtie-json-schema/bowtie/pull/970#issuecomment-1977491175

✅ Addressed extra padding issue.

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/cf3cef6b-343f-4fc6-8140-0d3235fcb638)

>Because I'm already looking at it (and sorry for the triple post) -- a dropdown menu with no label is very unusual! Probably add a title (to the whole thing) saying "Generate a Bowtie Badge", and then next to the dropdown menu a label on the left saying "Available Badges". https://github.com/bowtie-json-schema/bowtie/pull/970#issuecomment-1977494331

✅ Added title and associated label for the drop down menu.

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/a6f7d541-c715-4e4b-8513-0df1f96e21ec)

Julian commented:
Thanks. Your vertical alignment still isn't quite right, there's still extra padding on the left side:

![Screenshot 2024-03-06 at 9 08 20 AM](https://github.com/bowtie-json-schema/bowtie/assets/329822/84c89c1b-50d8-4408-bc2b-096ead331707)

You're probably still missing an `align-middle`.

Also please sort the dialects badges (by date, newest at the top).

sd1p reviewed (commented):
I was reading the decision on slack about `className` spaces issue. I tried to find a solution for this issue but didn't find anything helpful. So I corrected this problem on this PR.

Julian commented:
Great! Thanks! I'll have a closer look at the PR in the next few days. Appreciate the attention!

Julian commented on frontend/src/components/ImplementationReportView/BadgeFormats.tsx:
```diff
@@ -0,0 +1,65 @@
+export interface BadgeFormatOption {
+  type: string;
+  renderDisplay: (badgeURI: string) => JSX.Element;
+  generateCopyText: (badgeURI: string) => string;
```
What made you split this into its own property? It seems like it only adds the possibility for subtle bugs -- specifically us showing something to the user that's different than what we copy?

sd1p replied:
I thought about splitting `renderDisplay` and `generateCopyText` property specially and only to show `rSt` format. Other formats were easy to manage with a string property. Also made `BadgeFormatOptions` array homogeneous and developer friendly.

Julian replied:
What's special about rst?

sd1p replied:
To format `rSt` on render I used `<pre>` tag.  So any user can copy formatted badge with cursor also. I took the inspiration from https://shields.io/badges component.
![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/38009352-71e3-43e9-8a0a-6228e9db7dbb)

Julian replied:
I don't follow what's different about reST than any other format there.

Julian replied:
Yes but of course everything we show should be rendered the way we intend it to be rendered, not just reST.

Julian commented on frontend/src/components/ImplementationReportView/BadgeFormats.tsx:
```diff
@@ -0,0 +1,65 @@
+export const badgeFormatOptions: BadgeFormatOption[] = [
+  {
+    type: "URL",
+    renderDisplay: (badgeURI: string): JSX.Element => (
```
These `: JSX.Element`s don't look necessary, Typescript can figure them out from your `interface` definition.

Basically include the minimum amount of annotations to get it to compile :)

sd1p replied:
Thanks for the tip, I will make changes accordingly.

Julian commented on frontend/src/components/ImplementationReportView/BadgeFormats.tsx:
```diff
@@ -0,0 +1,65 @@
+    type: "Markdown",
+    renderDisplay: (badgeURI: string): JSX.Element => (
+      <span>
+        <pre className="pt-2 pb-2">![Static Badge]({badgeURI})</pre>
```
What do you mean here by `Static Badge`? The alt text of the badge should likely depend on which badge we're showing, and probably match the name of the badge.

sd1p replied:
This can be fixed by passing `prettyName` into renderDisplay and `generateCopyText` functions.

Julian commented on frontend/src/components/ImplementationReportView/EmbedBadges.tsx:
```diff
@@ -0,0 +1,173 @@
+const EmbedBadges: React.FC<{
+  implementation: Implementation;
+}> = ({ implementation }) => {
+  const [activeTab, setActiveTab] = useState("URL");
```
Let's focus the Markdown tab to start, as for better or worse that's the most common README format.

Julian commented on frontend/src/components/ImplementationReportView/EmbedBadges.tsx:
```diff
@@ -0,0 +1,173 @@
+  const [activeTab, setActiveTab] = useState("URL");
+  const [activeBadge, setActiveBadge] = useState("JSON Schema Versions");
+  const [badgeURI, setBadgeURI] = useState(
+    versionsBadgeFor(implementation).href(),
```
Convert to URI as late as possible (i.e. just before you actually render something), there's no reason to do it here I don't think.

Julian commented on frontend/src/components/ImplementationReportView/EmbedBadges.tsx:
```diff
@@ -0,0 +1,173 @@
+    versionsBadgeFor(implementation).href(),
+  );
+
+  const handleSelectBadge = (badgeName: string, URI: string): void => {
```
This `: void` is probably also not needed, check with TS.

Julian commented on frontend/src/components/ImplementationReportView/EmbedBadges.tsx:
```diff
@@ -0,0 +1,173 @@
+
+  const results = Object.entries(implementation.results).sort((a, b) => {
+    return (
+      Dialect.withName(b[0]).firstPublicationDate.getTime() -
```
Is the `.getTime` here actually needed? Javascript doesn't know how to just sort Dates themselves?

sd1p replied:
My linter was showing error `The left-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.` and `The right-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.` without using `.getTime()` method.

Julian replied:
Got it, ok, fair enough then.

Julian commented on frontend/src/components/ImplementationReportView/EmbedBadges.tsx:
```diff
@@ -0,0 +1,173 @@
+        type="button"
+        data-bs-toggle="dropdown"
+        data-bs-auto-close="outside"
+        style={{ width: "80px" }}
```
Please don't use `px` use some class here (and generally always so long as we are using Bootstrap).

Julian commented on frontend/src/components/ImplementationReportView/EmbedBadges.tsx:
```diff
@@ -0,0 +1,173 @@
+                data-bs-toggle="dropdown"
+                aria-expanded="false"
+              >
+                {activeBadge === "JSON Schema Versions"
```
You still have vestiges here of `if` statements checking what badge we're showing. Try to remove them!

Julian commented on frontend/src/components/ImplementationReportView/ImplementationReportView.tsx:
```diff
@@ -38,12 +39,20 @@ const ReportComponent: React.FC<{ implementation: Implementation }> = ({
+            <span>{implementation.name}</span>
+          </span>
+          <span>
+            {Object.entries(implementation.results).length !== 0 ? (
```
Does this ever really happen that it's empty?

sd1p replied:
> Does this ever really happen that it's empty?

I was not sure about it. I took precautions so that UI never breaks.

Julian replied:
It sounds like things are already broken if we ever have 0 results.

sd1p replied:
I'll take out the unnecessary safety precaution.

sd1p commented:
Thanks for reviewing my PR, I will implement the suggestions as soon as possible.

sd1p commented:
I have Implemented all the suggestions and improvements. Thanks for the review.

sd1p commented:
Thanks for refactoring the PR.  I will learn from the changes.

Julian commented:
If you care to have a look you might see if you can figure out where the right place to add `overflow-auto` to is to get the dropdown back to being fixed width and overflowing without hard coding the width and/or using `style`.

sd1p commented:
> If you care to have a look you might see if you can figure out where the right place to add `overflow-auto` to is to get the dropdown back to being fixed width and overflowing without hard coding the width and/or using `style`.

I will try to improve the CSS and remove inline CSS as much as possible. I will be having my midterm exams next week so will be pushing late but surely.

sd1p commented:
>If you care to have a look you might see if you can figure out where the right place to add overflow-auto to is to get the dropdown back to being fixed width and overflowing without hard coding the width and/or using style.

This commit doesn't fix this.

sd1p commented:
Changed the layout of the widget.

![image](https://github.com/bowtie-json-schema/bowtie/assets/43758514/fb292711-0009-4f0d-b62f-24ad601c8769)

Julian commented:
Thanks for your efforts here. There's more I'd like to tweak here but let's get this merged, I appreciate your persistence!

sd1p commented:
> Thanks for your efforts here. There's more I'd like to tweak here but let's get this merged, I appreciate your persistence!

Thank you for the continuous feedback and guidance.
